lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,566 @@
+           with_log(caller, "create pod", fn ->
+             create_pod(req, config, runtime_data, cluster_data.remote_port)
+           end),
+         _ <-
+           Phoenix.LiveView.send_update(
+             runtime.lv_pid,
+             LivebookWeb.SessionLive.K8sRuntimeComponent,
+             id: "runtime-config-k8s",
+             pod_name: pod_name
+           ),
+         {:ok, pod_ip} <-
+           with_pod_events(caller, "Waiting for pod to get ready", req, namespace, pod_name, fn ->
+             await_pod_ready(req, namespace, pod_name)
+           end),
+         _ <-
+           Phoenix.LiveView.send_update(
+             runtime.lv_pid,
+             LivebookWeb.SessionLive.K8sRuntimeComponent,
+             id: "runtime-config-k8s",
+             pod_name: nil
+           ),
+         child_node <- :"#{cluster_data.node_base}@#{pod_ip}",
+         :ok <-
+           with_log(caller, "start proxy", fn ->
+             k8s_forward_port(
+               kubeconfig,
+               context,
+               cluster_data,
+               pod_name,
+               namespace
+             )
+           end),
+         :ok <-
+           with_log(caller, "connect to node", fn ->
+             connect_loop(child_node, 40, 250)
+           end),
+         {:ok, primary_pid} <- fetch_runtime_info(child_node) do
+      primary_ref = Process.monitor(primary_pid)
+
+      server_pid =
+        with_log(caller, "initialize node", fn ->
+          initialize_node(child_node)
+        end)
+
+      send(primary_pid, :node_initialized)
+
+      runtime = %{runtime | node: child_node, server_pid: server_pid}
+      send(caller, {:runtime_connect_done, self(), {:ok, runtime}})
+
+      {:noreply, %{state | primary_ref: primary_ref}}
+    else
+      {:error, error} ->
+        send(caller, {:runtime_connect_done, self(), {:error, error}})
+
+        {:stop, :shutdown, state}
+    end
+  end
+
+  @impl true
+  def handle_info({:DOWN, ref, :process, _pid, _reason}, state) when ref == state.primary_ref do
+    {:stop, :shutdown, state}
+  end
+
+  def handle_info({port, _message}, state) when is_port(port) do
+    {:noreply, state}
+  end
+
+  defp get_free_port!() do
+    {:ok, socket} = :gen_tcp.listen(0, active: false, reuseaddr: true)
+    {:ok, port} = :inet.port(socket)
+    :gen_tcp.close(socket)
+    port
+  end
+
+  defp with_log(caller, name, fun) do
+    send(caller, {:runtime_connect_info, self(), name})
+
+    {microseconds, result} = :timer.tc(fun)
+    milliseconds = div(microseconds, 1000)
+
+    case result do
+      {:error, error} ->
+        Logger.debug("[K8s runtime] #{name} FAILED in #{milliseconds}ms, error: #{error}")
+
+      _ ->
+        Logger.debug("[K8s runtime] #{name} finished in #{milliseconds}ms")
+    end
+
+    result
+  end
+
+  defp with_pod_events(caller, name, req, namespace, pod_name, fun) do
+    with_log(caller, name, fn ->
+      runtime_pid = self()
+
+      event_watcher_pid =
+        spawn(fn ->
```

Comment:
The reason I didn't link the processes (correct me if I'm wrong): The current user might not have permissions to watch pods which is okay because this basically just adds more info. But if he doesn't have the permissions, the spawned process crashes and I didn't want it to pull the parent down as well. 

But now that I think about it, I can also just make the watcher a bit more fault tolerant...

Reply:
Yeah, instead of crashing in the watcher, we can match on error and just let the process finish normally, so it doesn't bring the caller down :)